bde/gaussian_crest.py

```diff
@@ -0,0 +1,282 @@
+
+
+    def optimize_molecule_mmff(self):
+        """ Embed a molecule in 3D space, optimizing a number of conformers and
+        selecting the most stable
+        """
+
+        mol = Chem.MolFromSmiles(self.smiles)
+        mol = Chem.rdmolops.AddHs(mol)
+
+        # If the molecule is a radical; add a hydrogen so MMFF converges to a
+        # reasonable structure
+        is_radical = False
+        radical_index = None
+        for i, atom in enumerate(mol.GetAtoms()):
+            if atom.GetNumRadicalElectrons() != 0:
+                is_radical = True
+                radical_index = i
+
+                atom.SetNumExplicitHs(atom.GetNumExplicitHs() + 1)
+                atom.SetNumRadicalElectrons(0)
+
+
+        # Use min < 3^n < max conformers, where n is the number of rotatable bonds
+        NumRotatableBonds = AllChem.CalcNumRotatableBonds(mol)
+        NumConformers = np.clip(3**NumRotatableBonds, self.min_conformers,
+                                self.max_conformers)
+
+        conformers = AllChem.EmbedMultipleConfs(
+            mol, numConfs=int(NumConformers), pruneRmsThresh=0.2, randomSeed=1,
+            useExpTorsionAnglePrefs=True, useBasicKnowledge=True)
+
+        def optimize_conformer(conformer):
+            prop = AllChem.MMFFGetMoleculeProperties(mol, mmffVariant="MMFF94s")
+            ff = AllChem.MMFFGetMoleculeForceField(mol, prop, confId=conformer)
+            ff.Minimize()
+            return float(ff.CalcEnergy())
+
+        assert conformers, "Conformer embedding failed"
+
+        if len(conformers) == 1:
+            logging.critical(
+                'Only 1 conformer for SMILES {}'.format(self.smiles))
+            most_stable_conformer = conformers[0]
+
+        else:
+            conformer_energies = np.array(
+                [optimize_conformer(conformer) for conformer in conformers])
+            most_stable_conformer = conformer_energies.argmin()
+
+        # If hydrogen was added; remove it before returning the final mol
+        if is_radical:
+            radical_atom = mol.GetAtomWithIdx(radical_index)
+            radical_atom.SetNumExplicitHs(int(radical_atom.GetNumExplicitHs()) - 1)
+            radical_atom.SetNumRadicalElectrons(1)
+
+        return mol, int(most_stable_conformer)
+
+
+    def run_crest(self, mol, confId, tmpdirname):
+        """ Given an rdkit.Mol object with an optimized, minimum energy conformer
+        ID, converting to .xyz and running crest in the scratch folder """
+
+        self.run_hex = uuid.uuid4().hex[:6]
+        self.xyz = tmpdirname + '/{0}_{1}.xyz'.format(self.cid, self.run_hex)
+        self.crest_out = tmpdirname + '/{0}_{1}.crest.out'.format(self.cid, self.run_hex)
+
+        #writing to xyzfile
+        rdkit.Chem.rdmolfiles.MolToXYZFile(mol,self.xyz,confId=confId)
+
+        #running calc in temporary TemporaryDirectory : NEEDS CHECK
+        with tempfile.TemporaryDirectory(dir=tmpdirname) as cresttmp:
```

Comment:
I'm not sure this is actually doing anything? you could put the other output in cresttmp, or chdir into cresttmp, but this is all already happening inside a temporary directory (tmpdirname) so the nested temp directory (i.e., what I did for gaussian) is probably not necessary